### ISSUE TITLE
[FEATURE]: Add option to invalidate .garden cache if garden version changes

### ISSUE THREAD

**Issue by Orzelius** [garden-io/garden#4136]
### Background / Motivation

Sometimes a garden version change behaves weirdly or breaks things in unpredictable ways if there's an existing cache created by an older garden version. Problems caused by this are very tricky to track down and often waste people's time.

### What should the user be able to do?

Specify in their project config if garden cache should automatically be cleared on a garden version change.

### Suggested Implementation(s)

Add a field like this to the project config

```yml
localCache:
  onGardenVersionChange: string  # either "clear" or "keep"
```

### How important is this feature for you/your team?

🌹 It’s a nice to have, but nice things are nice 🙂

**Comment by stefreak** [comment]
I think it might be worth considering making "clear" the default, or not exposing this option at all if we know it can cause unpredictable behaviour

**Comment by robkdev** [comment]
I'm the one who raised it.

I proposed that the version that generated the cache be added to the metadata.json file, then you can detect when it's generated by an older garden version, automatically clear the directory, and deploy from a clean slate.

It should be the default. It shouldn't require users having to know they have to clean the cache themselves or enable an obscure flag.

**Comment by Orzelius** [comment]
@robkdev Theoretically garden should work without having to clear the cache after a version change so that's why I'm hesitant to implement this.

For users who keep up with the edge releases or have massive projects it would be slow to regenerate the cache or redeploy from a clean slate each time.

**Comment by robkdev** [comment]
Not saying it should redeploy from clean slate each time, only when it detects version change.. or if you have well-defined schema models and can detect the difference at runtime.

I don't know whether it should work in theory, but I've seen it not work over half a dozen times now. It happens often enough that someone will update and have multiple deploys fail in a row before figuring out the problem or raising it, and that is a significant time waster. Having to deploy more than once takes many times longer than regenerating the cache.

**Comment by edvald** [comment]
That is true, and I'm not at all opposed to finding a good solution, just didn't quite care for the initial proposal .) 

I'd also be very keen to understand what exactly is causing the issue, so it would be helpful to see example logs, errors logs etc. from an occurrence of this.
